### PR TITLE
Make mobile app names unique by realm_id and platform

### PR DESCRIPTION
**Pull request by sethvargo**
Fixes GH-697

**Release Note**

```release-note
Make mobile app names unique by realm_id _and_ platform
```

### CODE REVIEW

File: pkg/database/mobile_app.go [hunk 1]
```diff
@@ -72,6 +72,23 @@ func (a *MobileApp) BeforeSave(tx *gorm.DB) error {
 		a.AddError("os", "is invalid")
 	}
 
+	// Ensure name/os combo is unique - there's still a database constraint that
+	// enforces this in case there's a race.
+	var existing MobileApp
+	if err := tx.Unscoped().
+		Model(&MobileApp{}).
+		Where("name = ?", a.Name).
```
Comment (whaught) [inline pkg/database/mobile_app.go:80]: Nit:
`Where("name = ? AND realm_id = ? AND os = ?", a.name, a.RealmID, a.OS)`
might be more SQL-y

Is this redundant with the constraint?

Comment (sethvargo) [reply pkg/database/mobile_app.go:80]: It's more SQL-y, but less Gorm-y. /shrug

No - this doesn't strictly guarantee that no duplicates occur. I just asserts that no duplicates occur at the time of the check. The constraint is a stronger promise, but that error is much harder to catch and throw back at the user. Doing this allows us to give the average user a nice "this already exists" response while still keeping the consistency guarantee of the uniqueness index.

**Comment by sethvargo** [comment]
@icco fixed the test failure

**Pull request merged**